### ISSUE TITLE
CFD simulation to predict pressure and velocity fields

### ISSUE THREAD

**Issue by asadabbas09** [rusty1s/pytorch_geometric#1747]
## ❓ Questions & Help

Thanks for sharing excellent work. 

I would like to do computational fluid dynamics simulation. (Something like https://www.youtube.com/watch?v=U38cKk-sxyY&ab_channel=AutodeskResearch) 

Paper Link: http://pub.ist.ac.at/~bbickel/downloads/2018_sigg_Learning3DAerodynamics.pdf 

**Dataset:** 

I have access to the dataset mentioned in the paper. 

It contains 

Inputs: shape S, incoming wind velocity V∞, mass density ρ, and viscosity μ. 

Output: drag coefficient and velocity fields and pressure values on the surface of the object.

I have read individual geometry files (*.vtk format), I can get access to `pos`, `tetra` and output pressure values on the surface.

My understanding is that for predicting pressure values on the surface and drag coefficient from input geometry I can convert it to point cloud and do regression on each node.

I'm not sure about the prediction of velocity fields around the object?

Would it be possible to use Geodesic operator (paper: https://arxiv.org/pdf/1802.04016.pdf) with GMMConv, to learn geometric features of the object and do convolution on the object surface?

Could you please elaborate if it's possible to work in this direction using GNNs and which one to use? And maybe give some pointers on how should I approach this problem. 

Thanks!

**Comment by rusty1s** [comment]
I'm absolutely no expert on this task, but this sounds like a perfect use-case for applying GNNs. If I understand you correctly, this is a node regression task, so one could simply start by applying a mesh/point cloud network and see how that performs. My guess is that additional input features might be crucial to obtain good performance, *e.g.*, normal vectors.

**Comment by asadabbas09** [comment]
Thanks for your quick response. I'm curious if I'm setting up the training pipeline correctly.

For predicting pressure values on each node (on mesh surface), I'm using `SplineConv` or `GMMConv`.

```class Net(torch.nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = SplineConv(1, 16, dim=3, kernel_size=5)
        self.conv2 = SplineConv(16, 32, dim=3, kernel_size=5)
        self.fc1 = torch.nn.Linear(32, 64)
        self.fc2 = torch.nn.Linear(64, 5334) #output number of nodes = 5334

    def forward(self, data):
        x = F.elu(self.conv1(data.x.float(), data.edge_index, data.edge_attr.float()))
        x = F.elu(self.conv2(x, data.edge_index, data.edge_attr.float()))
        #x = global_max_pool(x, data.batch)
        x = F.elu(self.fc1(x))
        x = F.dropout(x, training=self.training)
        x = self.fc2(x)
        return x
```

Few quick questions:

1. I'm confused about using `global_max_pool`. If I do use `global_max_pool`, it yields far lower training loss but does it makes sense for node regression task?
2. I'm using these transformations on input mesh dataset. Am I applying transformation correctly to feed it into `Spline/GMMConv` layer?
 ``` 
pre_transform = T.Compose([T.FaceToEdge(), T.Constant(value=1)])
transform=T.Spherical(). 
Data(edge_attr=[31992, 3], edge_index=[2, 31992], face=[3, 10664], pos=[5334, 3], x=[5334, 1]
```

3. I have also tried `graclus`, `max_pool`, `max_pool_x`, (similar to [MoNet.py](https://github.com/Farkor123/GeometricDeepLearning/blob/master/MoNet.py)), they also achieve similar performance. But not sure if they make sense for this task. 

Please let me know if you have any thoughts or suggestions on my approach.

**Comment by rusty1s** [comment]
1. No `global_max_pool` does not make sense for node regression tasks. It should also yield a shape mismatch error when computing the loss. You may want to try to increase the number of layers instead.
2. The transformations look correct, although I would use `T.Cartesian` instead of `T.Spherical` for slightly better performance (but this may be task-dependent).

**Comment by asadabbas09** [comment]
Thanks for your response, really appreciate it.

Is it possible to implement MeshGraphNets (Learning Mesh-Based Simulation with Graph Networks
https://arxiv.org/pdf/2010.03409.pdf) with PyTorch geometric, do you have any pointers? 

This work really looks interesting and something that we want to achieve.

**Comment by rusty1s** [comment]
The processor (Eq. (1)) can certainly be implemented in PyTorch Geometric, where you want to call `propagate` twice (for mesh and world edges, respectively). The Decoder then simply takes in latent node features to predict the next time state.

**Comment by scarpma** [comment]
Sorry for the dumb question, can I know why here
```python
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = SplineConv(1, 16, dim=3, kernel_size=5)
        self.conv2 = SplineConv(16, 32, dim=3, kernel_size=5)
        self.fc1 = torch.nn.Linear(32, 64)
        self.fc2 = torch.nn.Linear(64, 5334) #output number of nodes = 5334

    def forward(self, data):
        x = F.elu(self.conv1(data.x.float(), data.edge_index, data.edge_attr.float()))
        x = F.elu(self.conv2(x, data.edge_index, data.edge_attr.float()))
        #x = global_max_pool(x, data.batch)
        x = F.elu(self.fc1(x))
        x = F.dropout(x, training=self.training)
        x = self.fc2(x)
        return x
```
you are setting `self.fc2 = torch.nn.Linear(64, 5334) #output number of nodes = 5334` the number of nodes in the linear out_features ?

Setting `out_features=n_nodes` shouldn't its output be `n_nodes x n_nodes` (assuming a `batch_size=1`) ?

**Comment by scarpma** [comment]
Sorry @asadabbas09 and @rusty1s I really don't understand. Could you explain this to me ? Thank you very much !

**Comment by rusty1s** [comment]
Yes, the current output is `n_nodes x n_nodes`. You can change it dependent on your use-case.

**Comment by yhl48** [comment]
> I have access to the dataset mentioned in the paper.
@asadabbas09 would it be possible to share the dataset?
